cmd/snap/cmd_run.go

```diff
@@ -118,6 +118,40 @@ func getSnapInfo(snapName string, revision snap.Revision) (*snap.Info, error) {
 	return info, nil
 }
 
+func createOrUpdateUserDataSymlink(info *snap.Info, usr *user.User) error {
+	// create /home/snap/$SNAP_NAME/.lock file
+	lockFilePath := filepath.Join(info.HomeDirBase(usr.HomeDir), ".lock")
+	lockFd, err := syscall.Open(lockFilePath, syscall.O_CREAT, syscall.S_IRUSR|syscall.S_IWUSR)
+	if err != nil {
+		// TRANSLATORS: %q is the file path, %v the error message
+		return fmt.Errorf(i18n.G("cannot create lock file %q: %v"), lockFilePath, err)
+	}
+	err = syscall.Flock(lockFd, syscall.LOCK_EX)
```

Comment:
This can stall for unbound amount of time. I'd implement a safeguard like in snap-confine, where we setup a timer and abort if we cannot get the lock in some sane amount of time (e.g. 3 seconds).